GPUBindGroupDescriptor::externalTextureMatches is implemented incorrectly

mwyrzykowski opened the pull request:
#### ed52872fda631a8c522ac90d295838f964a88ac9
<pre>
GPUBindGroupDescriptor::externalTextureMatches is implemented incorrectly
<a href="https://bugs.webkit.org/show_bug.cgi?id=296062">https://bugs.webkit.org/show_bug.cgi?id=296062</a>
<a href="https://rdar.apple.com/155971582">rdar://155971582</a>

Reviewed by Tadeu Zagallo.

Auto-generated layouts can not be used interchangably, rather they
are fixed to a specific bind group layout and therefore bind group.

Prevent different auto-generated layouts from being intermixed, even
when their entries are otherwise identical.

This would previously always fail validation layer and is currently
preventing the drop down box on <a href="https://webgpu.github.io/webgpu-samples/?sample=videoUploading">https://webgpu.github.io/webgpu-samples/?sample=videoUploading</a>
from functioning.

* LayoutTests/fast/webgpu/regression/repro_296062-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_296062.html: Added.
Add regression test.

* Source/WebCore/Modules/WebGPU/GPUBindGroup.cpp:
(WebCore::GPUBindGroup::GPUBindGroup):
* Source/WebCore/Modules/WebGPU/GPUBindGroup.h:
(WebCore::GPUBindGroup::create):
(WebCore::GPUBindGroup::autogeneratedPipelineLayout const):
(WebCore::GPUBindGroup::GPUBindGroup): Deleted.
* Source/WebCore/Modules/WebGPU/GPUBindGroupLayout.h:
(WebCore::GPUBindGroupLayout::create):
(WebCore::GPUBindGroupLayout::autogeneratedPipelineLayout const):
(WebCore::GPUBindGroupLayout::GPUBindGroupLayout):
* Source/WebCore/Modules/WebGPU/GPUComputePipeline.cpp:
(WebCore::GPUComputePipeline::getBindGroupLayout):
* Source/WebCore/Modules/WebGPU/GPUComputePipeline.h:
(WebCore::GPUComputePipeline::create):
(WebCore::GPUComputePipeline::GPUComputePipeline):
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
* Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.h:
(WebCore::nextUniqueAutogeneratedPipelineIdentifier):
(WebCore::GPUPipelineDescriptorBase::uniqueAutogeneratedId const):
* Source/WebCore/Modules/WebGPU/GPURenderPipeline.cpp:
(WebCore::GPURenderPipeline::getBindGroupLayout):
* Source/WebCore/Modules/WebGPU/GPURenderPipeline.h:
(WebCore::GPURenderPipeline::create):
(WebCore::GPURenderPipeline::GPURenderPipeline):
Ensure auto-generated layouts don&apos;t allow mismatches.

Canonical link: <a href="https://commits.webkit.org/297771@main">https://commits.webkit.org/297771@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f34ad83403027047ed4109f20e24e359e6f97af5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112805 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23018 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119008 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63316 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114767 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33192 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41103 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85848 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115752 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26488 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101475 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66154 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25781 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19606 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62767 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95882 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19681 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122229 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39883 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29731 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94710 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40266 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97695 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94448 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24111 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39573 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17386 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39769 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45267 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39408 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42742 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41147 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebCore/Modules/WebGPU/GPUBindGroup.h:
```diff
@@ -47,14 +48,13 @@ class GPUBindGroup : public RefCounted<GPUBindGroup> {
     WebGPU::BindGroup& backing() { return m_backing; }
     const WebGPU::BindGroup& backing() const { return m_backing; }
     bool updateExternalTextures(GPUExternalTexture&);
+    uint64_t autogeneratedPipelineLayout() const { return m_autogeneratedPipelineLayout; }
```
Just a nit, but I find the name a bit confusing, since I'd expect this to return an actual layout object. If I understood correctly, isn't this a unique identifier for the pipeline?

mwyrzykowski replied:
oh yes, let me update that name

djg commented on Source/WebCore/Modules/WebGPU/GPUComputePipeline.cpp:
```diff
@@ -43,7 +43,7 @@ void GPUComputePipeline::setLabel(String&& label)
 Ref<GPUBindGroupLayout> GPUComputePipeline::getBindGroupLayout(uint32_t index)
 {
     // "A new GPUBindGroupLayout wrapper is returned each time"
```
Is this comment a TODO/FIXME? Or documenting the spec?

mwyrzykowski replied:
Documenting the spec

mwyrzykowski committed:
GPUBindGroupDescriptor::externalTextureMatches is implemented incorrectly
https://bugs.webkit.org/show_bug.cgi?id=296062
rdar://155971582

Reviewed by Tadeu Zagallo.

Auto-generated layouts can not be used interchangably, rather they
are fixed to a specific bind group layout and therefore bind group.

Prevent different auto-generated layouts from being intermixed, even
when their entries are otherwise identical.

This would previously always fail validation layer and is currently
preventing the drop down box on https://webgpu.github.io/webgpu-samples/?sample=videoUploading
from functioning.

* LayoutTests/fast/webgpu/regression/repro_296062-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_296062.html: Added.
Add regression test.

* Source/WebCore/Modules/WebGPU/GPUBindGroup.cpp:
(WebCore::GPUBindGroup::GPUBindGroup):
* Source/WebCore/Modules/WebGPU/GPUBindGroup.h:
(WebCore::GPUBindGroup::create):
(WebCore::GPUBindGroup::autogeneratedPipelineLayout const):
(WebCore::GPUBindGroup::GPUBindGroup): Deleted.
* Source/WebCore/Modules/WebGPU/GPUBindGroupLayout.h:
(WebCore::GPUBindGroupLayout::create):
(WebCore::GPUBindGroupLayout::autogeneratedPipelineLayout const):
(WebCore::GPUBindGroupLayout::GPUBindGroupLayout):
* Source/WebCore/Modules/WebGPU/GPUComputePipeline.cpp:
(WebCore::GPUComputePipeline::getBindGroupLayout):
* Source/WebCore/Modules/WebGPU/GPUComputePipeline.h:
(WebCore::GPUComputePipeline::create):
(WebCore::GPUComputePipeline::GPUComputePipeline):
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
* Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.h:
(WebCore::nextUniqueAutogeneratedPipelineIdentifier):
(WebCore::GPUPipelineDescriptorBase::uniqueAutogeneratedId const):
* Source/WebCore/Modules/WebGPU/GPURenderPipeline.cpp:
(WebCore::GPURenderPipeline::getBindGroupLayout):
* Source/WebCore/Modules/WebGPU/GPURenderPipeline.h:
(WebCore::GPURenderPipeline::create):
(WebCore::GPURenderPipeline::GPURenderPipeline):
Ensure auto-generated layouts don't allow mismatches.

Canonical link: https://commits.webkit.org/297771@main
